dbcon/mysql/ha_mcs_pushdown.cpp

```diff
@@ -792,71 +781,184 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
+        }
+
+        select_lex->update_used_tables();
 
         if (arena)
             thd->restore_active_arena(arena, &backup);
 
-        if (select_lex->handle_derived(thd->lex, DT_MERGE))
+#ifdef DEBUG_WALK_COND
+        if (conds)
         {
-            unsupported_feature = true;
-            handler->err_msg.assign("create_columnstore_select_handler(): \
-                Internal error occured in SL::handle_derived()");
+            conds->traverse_cond(cal_impl_if::debug_walk, NULL, Item::POSTFIX);
         }
+#endif
+    }
 
-        COND *conds = nullptr;
-        if (!unsupported_feature)
+    // Attempt to execute the query using the select handler.
+    // If query execution fails and columnstore_select_handler=AUTO,
+    // fallback to table mode.
+    // Skip execution for EXPLAIN queries
+    if (!thd->lex->describe)
+    {
+        // This is taken from JOIN::optimize()
+        join->fields= &select_lex->item_list;
+
+        // Instantiate handler::table, which is the place for the result set.
+        if (handler->prepare())
         {
-            // Rewrite once for PS
-	    // Refer to JOIN::optimize_inner() in sql/sql_select.cc
-	    // for details on the optimizations performed in this block.
-            if (select_lex->first_cond_optimization)
+            // check fallback
+            if (select_handler_mode == 2) // == 2 means columnstore_select_handler=AUTO
             {
-                create_explain_query_if_not_exists(thd->lex, thd->mem_root);
-                arena = thd->activate_stmt_arena_if_needed(&backup);
-                select_lex->first_cond_optimization= false;
-                conds = join->conds;
-                select_lex->where = conds;
+                push_warning(thd, Sql_condition::WARN_LEVEL_WARN, 9999,
+                    "MCS select_handler execution failed. Falling back to server execution");
+                restoreQueryState(handler);
+                delete handler;
+                return nullptr;
+            }
 
-                if (isPS)
-                {
-                    select_lex->prep_where = conds ? conds->copy_andor_structure(thd) : 0;
-                }
+            // error out
+            if (!thd->is_error())
+            {
+                my_printf_error(ER_INTERNAL_ERROR, "%s", MYF(0),
+                    "Error occured in handler->prepare()");
+            }
+
+            return handler;
+        }
+
+        //if (join->have_query_plan != JOIN::QEP_DELETED && join->build_explain())
+
+        // Prepare query execution
+        // This is taken from JOIN::exec_inner()
+        if (!select_lex->outer_select() &&                            // (1)
+            select_lex != select_lex->master_unit()->fake_select_lex) // (2)
+            thd->lex->set_limit_rows_examined();
+
+        if (!join->tables_list && (join->table_count || !select_lex->with_sum_func) &&
+            !select_lex->have_window_funcs())
+            return handler;
```

Comment:
Comment from Roman: check if `return nullptr;` is more appropriate here instead of `return handler;`. This also applies to other scenarios where we `return handler;`

Reply:
I have updated this part to `return nullptr;` if thd is not in an error state. Same is done for the `if (join->zero_result_cause)` branch.